crates/wit-component/src/encoding.rs

```diff
@@ -471,147 +619,769 @@ impl<'a> InterfaceEncoder<'a> {
-                } else {
-                    let mut encoded = match &ty.kind {
-                        TypeDefKind::Record(r) => self.encode_record(r)?,
-                        TypeDefKind::Variant(v) => self.encode_variant(v)?,
-                        TypeDefKind::List(ty) => {
-                            if matches!(ty, Type::Char) {
-                                InterfaceTypeRef::Primitive(PrimitiveInterfaceType::String)
+    fn export_type(
+        &mut self,
+        instance_ty: &mut Option<InstanceTypeEncoder<'a>>,
+        name: &'a str,
+        index: u32,
+    ) -> Result<()> {
+        if let Some(instance) = instance_ty.as_mut() {
+            return instance.export_type(name, index);
+        }
+
+        match self.exported.entry(name) {
+            Entry::Occupied(e) => {
+                if *e.get() != index {
+                    bail!("duplicate export `{}`", name)
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(index);
+                self.exports.export(name, ComponentExport::Type(index));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum RequiredOptions {
+    // No required options.
+    None,
+    // The encoding and into options are required.
+    Encoding,
+    // Only the into option is required.
+    Into,
+}
+
+impl RequiredOptions {
+    fn for_types<'a>(interface: &Interface, mut types: impl Iterator<Item = &'a Type>) -> Self {
+        match types.try_fold(Self::None, |acc, ty| {
+            match Self::for_type(interface, ty) {
+                Self::None => {
+                    // Don't update the accumulator
+                    Ok(acc)
+                }
+                Self::Encoding => {
+                    // If something requires the encoding option, bail early as an error
+                    Err(Self::Encoding)
+                }
+                Self::Into => {
+                    // Otherwise, update the accumulator (keep looking in case encoding is required)
+                    Ok(Self::Into)
+                }
+            }
+        }) {
+            Ok(o) | Err(o) => o,
+        }
+    }
+
+    fn for_type(interface: &Interface, ty: &Type) -> Self {
+        match ty {
+            Type::Id(id) => match &interface.types[*id].kind {
+                TypeDefKind::Record(r) => {
+                    Self::for_types(interface, r.fields.iter().map(|f| &f.ty))
+                }
+                TypeDefKind::Variant(v) => {
+                    Self::for_types(interface, v.cases.iter().filter_map(|c| c.ty.as_ref()))
+                }
+                TypeDefKind::List(t) => {
+                    if let Type::Char = t {
+                        return Self::Encoding;
+                    }
+
+                    match Self::for_type(interface, t) {
+                        Self::Encoding => Self::Encoding,
+                        _ => Self::Into,
+                    }
```

Comment:
One way to define this perhaps would be to define a `BitOr` implementation which is sort of a union up to the biggest thing, `Encoding`, so this would be:

```rust
Self::for_type(interface, t) | Self::Into
```

which would auto-propagate the `Encoding` if `t` requires `Encoding`

Reply:
I can add that.